### PR TITLE
BuildResidentialHPXML: detailed performance data arguments

### PR DESCRIPTION
**Pull request by joseph-robertson**
## Pull Request Description

Arguments added:
* `hvac_perf_data_capacity_type`
* `hvac_perf_data_heating_outdoor_temperatures`
* `hvac_perf_data_heating_min_speed_capacities`
* `hvac_perf_data_heating_max_speed_capacities`
* `hvac_perf_data_heating_min_speed_cops`
* `hvac_perf_data_heating_max_speed_cops`
* `hvac_perf_data_cooling_outdoor_temperatures`
* `hvac_perf_data_cooling_min_speed_capacities`
* `hvac_perf_data_cooling_max_speed_capacities`
* `hvac_perf_data_cooling_min_speed_cops`
* `hvac_perf_data_cooling_max_speed_cops`

## Checklist

PR Author: Check these when they're done. Not all may apply. ~~strikethrough~~ and check any that do not apply. 

PR Reviewer: Verify each has been completed.

- [ ] ~Schematron validator (`EPvalidator.xml`) has been updated~
- [ ] ~Sample files have been added/updated (`openstudio tasks.rb update_hpxmls`)~
- [x] Tests have been added/updated (e.g., `HPXMLtoOpenStudio/tests/test*.rb` and/or `workflow/tests/test*.rb`)
- [ ] ~Documentation has been updated~
- [x] Changelog has been updated
- [x] `openstudio tasks.rb update_measures` has been run
- [x] No unexpected changes to simulation results of sample files

### CODE REVIEW

File: BuildResidentialHPXML/measure.rb [hunk 1]
```diff
@@ -1382,6 +1420,74 @@ def arguments(model) # rubocop:disable Lint/UnusedMethodArgument
...
+    arg.setUnits('W/W')
+    args << arg
+
+    arg = OpenStudio::Measure::OSArgument.makeStringArgument('heat_pump_perf_data_heating_max_cops', false)
+    arg.setDisplayName('Heat Pump: Heating Performance Data Maximum Efficiency COP Values')
+    arg.setDescription('Maximum efficiency COP of heat pump heating detailed performance data. If multiple performance data points, use a comma-separated list.')
+    arg.setUnits('W/W')
+    args << arg
+
+    arg = OpenStudio::Measure::OSArgument.makeChoiceArgument('heat_pump_perf_data_cooling_cap_type', perf_data_capacity_type_choices, false)
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:1457]: No need for separate heating and cooling inputs, no one is going to want to use capacities for one and fractions for the other.

File: BuildResidentialHPXML/measure.rb [hunk 2]
```diff
@@ -1382,6 +1420,74 @@ def arguments(model) # rubocop:disable Lint/UnusedMethodArgument
     arg.setUnits('W')
     args << arg
 
+    arg = OpenStudio::Measure::OSArgument.makeChoiceArgument('heat_pump_perf_data_heating_cap_type', perf_data_capacity_type_choices, false)
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:1423]: I tend to think we should just have `heating_perf_data_foo` and `cooling_perf_data_foo` inputs, which means we don't need duplicate sets of cooling inputs. In general, the measure doesn't support having both an AC and a HP, though there might be some edge cases where it's possible...?

File: BuildResidentialHPXML/measure.rb [hunk 3]
```diff
@@ -1382,6 +1382,72 @@ def arguments(model) # rubocop:disable Lint/UnusedMethodArgument
...
+
+    arg = OpenStudio::Measure::OSArgument.makeChoiceArgument('perf_data_capacity_type', perf_data_capacity_type_choices, false)
+    arg.setDisplayName('Detailed Performance Data: Capacity Type')
+    arg.setDescription('Capacity type of detailed performance data.')
+    arg.setUnits('Capacity')
+    args << arg
+
+    arg = OpenStudio::Measure::OSArgument.makeStringArgument('heating_perf_data_outdoor_temperatures', false)
+    arg.setDisplayName('Detailed Performance Data: Heating Outdoor Temperatures')
+    arg.setDescription('Outdoor temperatures of heating detailed performance data. If multiple performance data points, use a comma-separated list.')
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:1397]: A few suggestions:
- For all these arguments, end the first sentence with "...... detailed performance data **if available**." to make it clearer that this is optional.
- For all these arguments, add a sentence "Applies only to variable-speed air-source HVAC systems (central air conditioners, mini-split air conditioners, air-to-air heat pumps, and mini-split heat pumps)."
- For `heating_perf_data_outdoor_temperatures`, add a sentence: "One of the outdoor temperatures must be 47 deg-F."
- For `cooling_perf_data_outdoor_temperatures`, add a sentence: "One of the outdoor temperatures must be 95 deg-F."

File: BuildResidentialHPXML/measure.rb [hunk 4]
```diff
@@ -1382,6 +1382,72 @@ def arguments(model) # rubocop:disable Lint/UnusedMethodArgument
...
+    arg.setUnits('deg-F')
+    args << arg
+
+    arg = OpenStudio::Measure::OSArgument.makeStringArgument('heating_perf_data_minimum_capacities', false)
+    arg.setDisplayName('Detailed Performance Data: Heating Minimum Capacities')
+    arg.setDescription('Minimum capacities of heating detailed performance data. If multiple performance data points, use a comma-separated list.')
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:1403]: Need to be careful with terminology, _especially for COPs_. What you call "minimum COPs" is really "COPs at minimum speed" and they will have **larger** values than the "maximum COPs", which is really "COPs at maximum speed".

So make changes to all arguments as follows:
- "Heating Minimum Foos" => Heating Foos at Minimum Speed
"Minimum foos of heating detailed performance data." => "Foos at minimum speed of heating detailed performance data."

File: BuildResidentialHPXML/measure.rb [hunk 5]
```diff
@@ -1382,6 +1382,72 @@ def arguments(model) # rubocop:disable Lint/UnusedMethodArgument
...
+    args << arg
+
+    arg = OpenStudio::Measure::OSArgument.makeStringArgument('heating_perf_data_maximum_capacities', false)
+    arg.setDisplayName('Detailed Performance Data: Heating Maximum Capacities')
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:1407]: All of the capacity arguments should have units of "Btuh/hr or Frac".

**Review by shorowit — COMMENTED** [review]
Great implementation. I mostly have a bunch of comments around clarifying argument descriptions.

File: BuildResidentialHPXML/measure.rb (hunk 3 above)
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:1397]: "If multiple performance data points, use a comma-separated list."

For all of these arguments, multiple data points are _required_. So perhaps the language should be something like: "At least two performance data points are required using a comma-separated list."

File: BuildResidentialHPXML/measure.rb [hunk 6]
```diff
@@ -3395,6 +3461,44 @@ def argument_errors(args)
     error = args[:rim_joist_assembly_r].is_initialized && !args[:geometry_rim_joist_height].is_initialized
     errors << 'Specified a rim joist assembly R-value but no rim joist height.' if error
 
+    heating_perf_data_args_initialized = [args[:heating_perf_data_outdoor_temperatures].is_initialized,
+                                          args[:heating_perf_data_minimum_capacities].is_initialized,
+                                          args[:heating_perf_data_maximum_capacities].is_initialized,
+                                          args[:heating_perf_data_minimum_cops].is_initialized,
+                                          args[:heating_perf_data_maximum_cops].is_initialized]
+    error = (heating_perf_data_args_initialized.uniq.size != 1)
+    errors << 'Did not specify all required heating detailed performance data arguments.' if error
```
[the remark this answers is not in the document]
Comment (joseph-robertson) [reply BuildResidentialHPXML/measure.rb]: Separate PR to cover this?

[the remark this answers is not in the document]
Comment (shorowit) [reply BuildResidentialHPXML/measure.rb]: That's fine.

**Review by shorowit — APPROVED** [review]
Looks good to me. If you don't have a problem with my latest commit, I can merge this after the CI comes back happy.

**Pull request merged**